cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStateWriter.java

```diff
@@ -95,6 +95,16 @@ public void running(ProgramRunId programRunId, @Nullable String twillRunId) {
     programStatePublisher.publish(Notification.Type.PROGRAM_STATUS, properties.build());
   }
 
+  @Override
+  public void stop(ProgramRunId programRunId, long gracefulShutdownPeriod) {
+    ImmutableMap<String, String> properties = ImmutableMap.<String, String>builder()
```

Comment:
You probably want to guard against gracefulShutdownPeriod < 0.

Reply:
Should this logic be closer to the rest layer where we validate the graceful timeout and throw an error if it is invalid?

Reply:
Will have the validation logic on the handler side in the future PR as discussed offline